pkg/client/apiutil/restmapper_test.go

```
@@ -564,6 +550,131 @@ func TestLazyRestMapperProvider(t *testing.T) {
+
+		// We update the CRD to only have v1 version.
+		g.Expect(c.Get(ctx, types.NamespacedName{Name: crdName}, crd)).To(gmg.Succeed())
+		var v1 apiextensionsv1.CustomResourceDefinitionVersion
+		for i, version := range crd.Spec.Versions {
+			if version.Name == "v1" {
+				crd.Spec.Versions[i].Storage = true
+				v1 = version
+				v1.Storage = true
+			}
+		}
+		crd.Spec.Versions = []apiextensionsv1.CustomResourceDefinitionVersion{v1}
+		g.Expect(c.Update(ctx, crd)).To(gmg.Succeed())
+
+		// We wait until v2 is not available anymore.
+		g.Eventually(func(g gmg.Gomega) {
+			_, err = discClient.ServerResourcesForGroupVersion(group + "/v2")
+			g.Expect(apierrors.IsNotFound(err)).To(gmg.BeTrue(), "v2 should not be available anymore")
+		}).Should(gmg.Succeed())
+
+		// Although v2 is not available anymore, the cache is not invalidated yet so it should return a mapping.
+		_, err = lazyRestMapper.RESTMapping(schema.GroupKind{Group: group, Kind: kind}, "v2")
+		g.Expect(err).NotTo(gmg.HaveOccurred())
+		g.Expect(crt.GetRequestCount()).To(gmg.Equal(0))
+
+		// We request Limo, which is not in the mapper because it doesn't exist.
+		// This will trigger a reload of the lazy mapper cache.
+		// Reloading the cache will read v2 again and since it's not available anymore, it should invalidate the cache.
+		// 	#1: GET https://host/apis/inventory.example.com/v1
+		// 	#2: GET https://host/apis/inventory.example.com/v2
+		_, err = lazyRestMapper.RESTMapping(schema.GroupKind{Group: group, Kind: "Limo"})
+		g.Expect(err).To(beNoMatchError())
+		g.Expect(crt.GetRequestCount()).To(gmg.Equal(2))
+		crt.Reset()
+
+		// Now we request v2 again and it should return an error since the cache was invalidated.
+		// 	#1: GET https://host/apis/inventory.example.com/v2
+		_, err = lazyRestMapper.RESTMapping(schema.GroupKind{Group: group, Kind: kind}, "v2")
+		g.Expect(err).To(beNoMatchError())
+		g.Expect(crt.GetRequestCount()).To(gmg.Equal(1))
```

Should we do one more call to verify calling RESTMappings without versions set would work and only return v1?

(I think this is the case that we hit through apiutil.IsGVKNamespaced)

---

That wouldn't happen with this version of CR since not found errors were ignored already. Added an extra call to make it explicit jic.

---

Good point